runtime/compiler/control/CompilationThread.cpp

```diff
@@ -837,46 +837,48 @@ TR::CompilationInfoPerThread::relocateThunks()
    TR_J9VMBase *fe = _vm;
    for (auto p : _thunksToBeRelocated)
       {
-      void *thunk = (U_8 *)p.first - (U_8 *)TR::compInfoPT->reloRuntime()->aotMethodHeaderEntry()->compileMethodCodeStartPC + (U_8 *)TR::compInfoPT->reloRuntime()->newMethodCodeStart();
-      std::string signature = p.second;
-      void *vmHelper = j9ThunkVMHelperFromSignature(fe->_jitConfig, signature.size(), &signature[0]);
-      TR::compInfoPT->reloRuntime()->reloTarget()->performThunkRelocation((uint8_t*) thunk, (UDATA)vmHelper);
-      // Always need to call base version of setJ2IThunk, even in AOT mode
-      fe->TR_J9VMBase::setJ2IThunk(&signature[0], signature.size(), thunk, TR::comp());
-      if (_vm->_compInfoPT->getCompilation()->compileRelocatableCode())
-         // For AOT compilation, also need to call TR_J9SharedCacheVM version of the method
-         _vm->setJ2IThunk(&signature[0], signature.size(), thunk, TR::comp());
-      }
-   _thunksToBeRelocated.clear();
-   for (TR_J2IThunk *thunk : _invokeExactThunksToBeRelocated)
-      {
-      TR_J2IThunk *realThunk = (TR_J2IThunk *)((U_8 *)thunk - (U_8 *)TR::compInfoPT->reloRuntime()->aotMethodHeaderEntry()->compileMethodCodeStartPC + (U_8 *)TR::compInfoPT->reloRuntime()->newMethodCodeStart());
-      char *signature = realThunk->terseSignature();
-      void *vmHelper = j9ThunkInvokeExactHelperFromTerseSignature(fe->_jitConfig, strlen(signature), signature);
-      *(UDATA *)(realThunk->entryPoint() + 2) = (UDATA) vmHelper; // JITaaS TODO: This is amd64 specific
-      fe->setInvokeExactJ2IThunk(realThunk, TR::comp());
-      }
-   _invokeExactThunksToBeRelocated.clear();
-   }
+      std::string &serializedThunk = p.first;
+      std::string &signature = p.second;
 
-void
-TR::CompilationInfoPerThread::persistThunksToSCC(const J9JITDataCacheHeader *cacheEntry, uint8_t *existingCode)
-   {
-   TR_ASSERT(_compInfo.getPersistentInfo()->getRemoteCompilationMode() == JITServer::CLIENT, "Should be called in JITServer client mode only");
-   TR_ASSERT(_vm->isAOT_DEPRECATED_DO_NOT_USE(), "Should be for AOT mode only");
+      uint8_t *coldCode;
+      TR::CodeCache *codeCache = reloRuntime()->codeCache();
```

Comment:
Yes, we have codeCache reserved at that point, so allocation is possible.
I've seen a test where we actually do run out of space and cannot allocate here, I don't know whether we could use some other cold cache, or cold cache here.
I just did the same thing that `TR_RelocationRecordThunks::relocateAndRegisterThunks` does. They use warm code cache there.

Reply:
We used to allocate thunks in the cold area of the code cache on the server, since thunks are never actually run there.
I changed that code in this PR to allocate thunks on the heap instead, since we immediately send thunks to the client now.